external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CachedKafkaProducerSuite.scala

```diff
@@ -18,60 +18,208 @@
+
+    def release(producer: CachedKafkaProducer): Unit = {
+      if (producer != null) {
+        CachedKafkaProducer.release(producer, Random.nextBoolean())
+        if (producer.getInUseCount > 0) {
+          assert(!producer.isClosed, "Should not close an inuse producer.")
+        }
+      }
+    }
+    val threadPool = Executors.newFixedThreadPool(numThreads)
+    try {
+      val futuresAcquire = (1 to 10 * numConcurrentProducers).map { i =>
+        threadPool.submit(new Runnable {
+          override def run(): Unit = {
+            acquire(i % numConcurrentProducers + 1)
+          }
+        })
+      }
+      val futuresRelease = (1 to 10 * numConcurrentProducers).map { i =>
+        val cachedKafkaProducer = toBeReleasedQueue.poll()
+        // 2x release should not corrupt the state of cache.
+        (1 to 2).map { j =>
+          threadPool.submit(new Runnable {
+            override def run(): Unit = {
+              release(cachedKafkaProducer)
+            }
+          })
+        }
+      }
+      futuresAcquire.foreach(_.get(1, TimeUnit.MINUTES))
+      futuresRelease.flatten.foreach(_.get(1, TimeUnit.MINUTES))
+    } finally {
+      threadPool.shutdown()
+      CachedKafkaProducer.clear()
+    }
+  }
+
+  /*
+   * The following stress suite will cause frequent eviction of kafka producers from
+   * the guava cache. Since these producers remain in use, because they are used by
+   * multiple tasks, they stay in close queue till they are released finally. This test
+   * will cause new tasks to use fresh instance of kafka producers and as a result it
+   * simulates a stress situation, where multiple producers are requested from CachedKafkaProducer
+   * and at the same time there will be multiple releases. It is supposed to catch a race
+   * condition if any, due to multiple threads requesting and releasing producers.
+   */
+  test("Single source and multiple kafka sink with 2ms cache timeout.") {
+
+    val df = spark.readStream
+      .format("rate")
+      .option("numPartitions", "100")
+      .option("rowsPerSecond", "200")
+      .load()
+      .selectExpr("CAST(timestamp AS STRING) key", "CAST(value AS STRING) value")
+
+    val checkpointDir = Utils.createTempDir()
+    val topic = newTopic()
+    testUtils.createTopic(topic, 100)
+    failAfter(streamingTimeout) {
+      val queries = for (i <- 1 to 10) yield {
+        df.writeStream
+          .format("kafka")
+          .option("checkpointLocation", checkpointDir.getCanonicalPath + i)
+          .option("kafka.bootstrap.servers", testUtils.brokerAddress)
+          // to make it create 5 unique producers.
+          .option("kafka.max.block.ms", s"100${i % 5}")
+          .option("topic", topic)
+          .trigger(Trigger.Continuous(500))
+          .queryName(s"kafkaStream$i")
+          .start()
+      }
+      Thread.sleep(15000)
+
+      queries.foreach { q =>
+        assert(q.exception.isEmpty, "None of the queries should fail.")
+        q.stop()
```

Comment:
Queries won't be stopped if `q.exception.isEmpty` condition not fulfilled, will they?

Reply:
There are existing tests, where querry.stop() occurs after the assert statement. Do you want me to ensure that q.stop is called inspite of a failure?

Reply:
This case is different because if the first query gives exception the rest of the queries may run properly and has to be stopped.